docs: extract sections in how to write integration tests to their own howto guide (#2662)
canonical/operator: docs/howto/manage-relations.md

To integrate with another charm, or with itself (to communicate with other units of the same charm), declare the required and optional relations in your charm's `charmcraft.yaml` file.

```{note}

**If you're using an existing interface:**

Make sure to consult {ref}`manage-interfaces` and {external+charmlibs:doc}`how-to/manage-libraries` for guidance about how to implement them correctly.

**If you're defining a new interface:**

> Make sure to add your interface to [the `charmlibs` repository](https://github.com/canonical/charmlibs). See {external+charmlibs:doc}`how-to/design-relation-interfaces`, {external+charmlibs:doc}`how-to/python-package`.

Reviewer: Maybe we should add some sort of "If this is an interface other charms should use" caveat here? I'm not sure we are 100% sorted on exactly which interfaces should/shouldn't go into charmlibs.

**If you're using an existing interface:**

> Make sure to consult {external+charmlibs:doc}`how-to/design-relation-interfaces` for guidance about how to implement them correctly.

Author: Drive by fix. https://github.com/canonical/charm-relation-interfaces is now archived.

Reviewer 2: The linked doc ("How to design relation interfaces") is for people developing an interface. But the first part of the note look like it's for people _using_ an interface. I think we should link to [https://canonical.com/juju/docs/ops/latest/howto/manage-libraries/](https://canonical.com/juju/docs/ops/latest/howto/manage-libraries/) and [https://canonical.com/juju/docs/charmlibs/how-to/manage-libraries/](https://canonical.com/juju/docs/charmlibs/how-to/manage-libraries/) instead.

The link to "How to design relation interfaces" could go in the second part of the note, before "How to distribute charm libraries".

While we're here, I recommend deescalating this from a Caution note to a normal note.

Author: Thanks for this. 

I think I confused myself over `how to implement them correctly`. I read it as "implementing the library itself", while it should be understand as "implementing the charm code to use the library" :thinking: 

I will follow your suggestion.

For example:

```python

> class SMTPProviderUnitData(pydantic.BaseModel):

Reviewer: Nice catch :)

`remote`:

- `Relation.remote_units_data` becomes `SubordinateRelation.remote_unit_data`

>   taking a single `dict[str, str]`. The remote unit ID can be provided as a

Reviewer: Here too :)

relation.remote_unit_name  # 'zookeeper/42'
```

(manage-relations-integration-tests)=
### Write integration tests

> See first: {ref}`write-integration-tests-for-a-charm`

> To verify that a charm behaves correctly when integrated with another in a real Juju environment, write an integration test with `jubilant` that deploys another application and relates your charm to it.

Reviewer 2: I recommend dropping this or shortening it. Something like:

```suggestion
Use Jubilant to verify that your charm behaves correctly when integrated with another charm:
```

> See first: {ref}`write-integration-tests-for-a-charm`

To verify that charm behaves correctly when integrated with another in a real Juju environment, write an integration test with `jubilant` that deploys another application and relates your charm to it.
To verify that a charm behaves correctly when integrated with another in a real Juju environment, write an integration test with `jubilant` that deploys another application and relates your charm to it.

```python
# This assumes that your integration tests already include the standard
# build and deploy test.

def test_active_with_another_app(juju: jubilant.Juju):
def test_active_with_another_app(charm: pathlib.Path, juju: jubilant.Juju):
    juju.deploy('another-app')
    juju.integrate('your-app:endpoint', 'another-app:endpoint')

> 

Author: This is a style change to keep it consistent with other locations, where there is no new line in between the `juju.*` invocations.

See more: [](jubilant.Juju.integrate)

<!-- valkey-operator example is probably too complicated -->
Examples for testing relations with Jubilant: [`valkey-operator`](https://github.com/canonical/valkey-operator/blob/9/edge/tests/integration/clients/test_client_relation.py), [`kafka-k8s-operator`](https://github.com/canonical/kafka-k8s-operator/blob/main/tests/integration/test_provider_v1.py).

> This test (and subsequent tests) don't need to depend on the `charm` fixture. However, it's helpful for each test to depend on `charm`, so that each test fails immediately if a `.charm` file isn't available.

Author: This paragraph was originally from https://canonical.com/juju/docs/ops/latest/howto/write-integration-tests-for-a-charm/#test-a-relation. As we moved those tests into their own Howto guides, I wonder if we should either:
1. Duplicate this paragraph in those Howto guides (`Manage relation`, `Manage resources`, and `Manage configuration`) OR
1. Remove this paragraph and the `charm` fixture completely from the relevant tests.

Since we encourage using the `charm` fixture to fail early, I am okay with the first solution (and remove the `and subsequent tests` phrase).

Reviewer 2: I don't like having to explain this everywhere, but I think that including `charm` in every test is the most practical/uncomplicated way to ensure fast failure. So we ought to explain it, since it's not intuitive.

I think it's good to keep the explanation as concise as we can. Maybe something like:

> We recommend including the `charm` fixture (even though it's not used) so that the test fails immediately if a `.charm` file isn't available.

I agree that we might as well mention this in the other how-to guides too.
